src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp

```diff
@@ -402,6 +403,96 @@ struct KVAxesPosition {
+            if (node_to_output.count(matmul_add)) {
+                matched_node_last_op = node_to_output.at(matmul_add).get_node_shared_ptr();
+            } else if (node_to_output.count(matmul_transpose)) {
+                matched_node_last_op = node_to_output.at(matmul_transpose).get_node_shared_ptr();
+            } else if (node_to_output.count(matmul_convert)) {
+                matched_node_last_op = node_to_output.at(matmul_convert).get_node_shared_ptr();
+            } else if (node_to_output.count(matmul_multiply)) {
+                matched_node_last_op = node_to_output.at(matmul_multiply).get_node_shared_ptr();
+            } else {
+                matched_node_last_op = matched_node_matmul;
+            }
+            auto matched_node_result = node_to_output.at(res).get_node_shared_ptr();
+
+            auto matched_matmul = std::static_pointer_cast<ov::op::v0::MatMul>(matched_node_matmul);
+            auto matched_result = std::static_pointer_cast<ov::op::v0::Result>(matched_node_result);
+
+            // Cut point:
+            auto matmul_first_source = matched_matmul->input(0).get_source_output();
+
+            // Cut original model:
+            matched_result->input(0).replace_source_output(matmul_first_source);
+            // FIXME: Somehow for KVCache model result output gets renamed in
+            //        ICompiledModel::ICompiledModel().
+            //        As a WA, setting the same name to output from MatMul
+            //        avoids the issue.
+            matmul_first_source.set_names({"output_embed"});
+            matched_result->output(0).set_names({"output_embed"});
+            matched_result->validate_and_infer_types();
+
+            // Create an additional model after cut point:
+            auto new_param = std::make_shared<ov::op::v0::Parameter>(matmul_first_source.get_element_type(),
+                                                                     matmul_first_source.get_partial_shape());
+            new_param->output(0).add_names({"output_embed"});
+            matched_matmul->input(0).replace_source_output(new_param);
+            auto new_result = std::make_shared<ov::op::v0::Result>(matched_node_last_op);
+            tail_vocab_mm_ref.get() = std::make_shared<ov::Model>(ov::OutputVector{new_result->output(0)},
+                                                                  ov::ParameterVector{new_param},
+                                                                  "TailVocabMatmul");
+
+            return true;
+        };
+        register_matcher(std::make_shared<opp::Matcher>(res, "CutTailMatMul"), std::move(callback));
+    }
+};
+
+namespace {
+std::shared_ptr<ov::Model> cut_tail_matmul(std::shared_ptr<ov::Model>& model) {
+    ov::pass::GraphRewrite rewr;
+    std::shared_ptr<ov::Model> tail_vocab_mm = nullptr;
+    std::reference_wrapper<std::shared_ptr<ov::Model>> tail_vocab_mm_ref(tail_vocab_mm);
```

Comment:
ES: use just `std::shared_pointer`

Reply:
Unfortunately as `ov::Model` can't be created with default constructor, we can't create dummy model to rewrite it later.

Reply:
Fixed, thanks!!